skll/learner.py

```diff
@@ -1465,56 +1533,55 @@ def train(self, examples, param_grid=None, grid_search_folds=3,
-                    # if all of the labels are strings, only class 1 objectives
-                    # are valid (with a classifier).
-                    raise ValueError("{} is not a valid grid objective "
-                                     "function for the {} learner with string "
-                                     "labels".format(grid_objective,
-                                                     self._model_type.__name__))
-
-                elif len(set(examples.labels)) == 2:
-                    # If there are two labels, class 3 objectives are valid for
-                    # classifiers regardless of the type of the label.
-                    if grid_objective not in _BINARY_CLASS_OBJ_FUNCS:
-                        raise ValueError("{} is not a valid grid objective "
-                                         "function for the {} learner with "
-                                         "binary labels"
-                                         .format(grid_objective,
-                                                 self._model_type.__name__))
-                elif grid_objective in _REGRESSION_ONLY_OBJ_FUNCS:
-                    # simple backoff check for mixed-type labels
-                    raise ValueError("{} is not a valid grid objective "
-                                     "function for the {} learner"
-                                     .format(grid_objective,
-                                             self._model_type.__name__))
+
+            # get the list of objectives that are acceptable in the current
+            # prediction scenario and raise an exception if the current
+            # objective is not in this allowed list
+            label_type = examples.labels.dtype.type
+            if estimator_type == 'classifier':
+                sorted_unique_labels = np.unique(examples.labels)
+                allowed_objectives = _get_acceptable_classification_metrics(sorted_unique_labels)
+            else:
+                allowed_objectives = _get_acceptable_regression_metrics()
+
+            if grid_objective not in allowed_objectives:
+                raise ValueError("'{}' is not a valid objective "
+                                 "function for {} with "
+                                 "labels of type {}.".format(grid_objective,
+                                                             self._model_type.__name__,
+                                                             label_type.__name__))
+
+            # If we're using a correlation metric for doing binary
+            # classification and probability is set to true, we assume
+            # that the user actually wants the `_with_probabilities`
+            # version of the metric
+            if (grid_objective in _CORRELATION_METRICS and
+                    estimator_type == 'classifier' and
+                    self.probability):
+                self.logger.info('You specified "{}" as the objective with '
```

Comment:
The value that is supposed to be inserted in `{}` is never actually supplied (via `.format(...)`.